src/transform/src/reduction_pushdown.rs

```diff
@@ -81,8 +85,581 @@ impl ReductionPushdown {
+            } else {
+                self.evaluated_expressions.insert(
+                    self.old_join_mapper.map_expr_to_global(group_key, idx),
+                    MirScalarExpr::Column(self.next_new_join_column),
+                );
+            }
+            self.next_new_join_column += 1;
+        }
+        for (agg_num, agg) in inner_aggs {
+            let agg_lookup_key = self.old_join_mapper.total_columns() + agg_num;
+            let (outer_agg_func, downcast) = agg.func.outer_agg();
+            // calculate aggs to add to outer reduce
+            let outer_agg = AggregateExpr {
+                func: outer_agg_func,
+                expr: MirScalarExpr::Column(agg_lookup_key),
+                distinct: agg.distinct,
+            };
+            if let Some(downcast) = downcast {
+                self.downcast_partial_reduce.push(MirScalarExpr::CallUnary {
+                    func: downcast,
+                    expr: Box::new(MirScalarExpr::Column(self.outer_group_key.len() + agg_num)),
+                });
+            }
+            self.outer_aggs.push((agg_num, outer_agg));
+            self.permutation
+                .insert(agg_lookup_key, self.next_new_join_column);
+            self.next_new_join_column += 1;
+        }
+        self.pushdown_succeeded = true;
+    }
+}
+
+/// Construct a graph showing, for each input, which other inputs it is joined
+/// to and on which join keys.
+/// Example:
+/// For the query
+/// """
+/// select *
+/// from foo, bar, baz
+/// where foo.a = bar.b
+/// and foo.c = bar.d
+/// and bar.e = baz.f
+/// """
+/// The corresponding graph will be
+/// "foo" -> ["bar" -> ["a", "c"]]
+/// "bar" -> ["foo" -> ["b", "d"], "baz" -> ["e"]]
+/// "baz" -> ["baz" -> ["f"]]
+fn create_join_graph(
+    old_join_mapper: &JoinInputMapper,
+    equivalences: &[Vec<MirScalarExpr>],
+) -> Option<Vec<HashMap<usize, Vec<MirScalarExpr>>>> {
+    let mut join_keys_by_input = vec![HashMap::new(); old_join_mapper.total_inputs()];
+    for equivalence in equivalences.iter() {
+        // track all inputs referenced in the equivalence class
+        let mut inputs_in_equivalence = Vec::new();
+        // localize each expression in equivalence to its respective input
+        let mut expr_with_inputs = Vec::new();
+        for expr in equivalence.iter() {
+            let mut expr_inputs = old_join_mapper.lookup_inputs(expr);
+            if let Some(expr_input) = expr_inputs.next() {
+                if expr_inputs.next().is_none() {
+                    expr_with_inputs.push((expr_input, expr.clone()));
+                    inputs_in_equivalence.push(expr_input);
+                } else {
+                    // there is an expression in a join equivalence class that
+                    // spans multiple inputs. For now, skip reduction pushdown
+                    // in this case.
+                    return None;
+                }
+            }
+        }
+        if inputs_in_equivalence.len() > 1 {
```

Comment:
why does this have to be greater than 1? it seems like that would rule out cases like `select * from foo where foo.a = foo.b` is that right? I'm just trying to double check my understanding here. Maybe consider a comment above the if statement?

Reply:
Looking at the crash that @philip-stoev reminded me that we often push filters down where a column is equal to a constant into the join equivalences.
`Join %0 %1 (= #0 #16) (= #17 196) (= #11 1994-09-23) (= #3 1)`, so we definitely need it to rule out cases like
`select * from foo, bar where foo.a = 1`. It may also be used to rule out cases like `select * from foo, bar where foo.a = foo.b`.

@frankmcsherry Perhaps we should revert this behavior since indexes to speed up filtering now happens at the render 
layer?

Reply:
well regardless of whether we revert this behavior, indeed this code is problematic because if `inputs_in_equivalence.len() <= 1`, it is a single-input filter, which should be pushed down when reduction gets pushed down. Currently, the single-input filter is ignored, which is what is causing the crash.

Aka
```
Reduce {
  Filter {
    Join {foo, bar}
    predicates: foo.a = 1
  }
  agg: something
}
```

needs to become

```
Reduce {
  Join {
    Reduce{
      Filter {foo, predicates: foo.a = 1}, 
      agg: something
    },
    bar
  }
agg: something
}
```